### DISCUSSION TITLE
Custom Tokenizer: separating words with ":" in between

### DISCUSSION THREAD

**Discussion by farrandi** [explosion/spaCy#9203] [Help: Coding & Implementations] [answered]
I have a problem where for example a word "abc:123" would be recognized as 1 token by spacy.
I tried customizing the tokenizer infix:
```
infixes = default stuff + [r"(?<=[{a}0-9])[:;()/=](?=[{a}0-9])".format(a=ALPHA)] # added this to separate 'obs:30'
infix_re = compile_infix_regex(infixes)
nlp.tokenizer.infix_finditer = infix_re.finditer
```

This does indeed separate "abc:123" into `[abc, :, 123]`. However this does not work when the digit following is only 1 digit:
"abc:1" would give: `[abc, :1]`

((weirdly "abc:a" would give `[abc, :, a]`
I was wondering why this is the case? and how I could make it so that "abc:1 would be separated into 3 tokens.

---------------------------------------------------------------------------------------------------------------------------

I modified the tokenizer a lot, here is the full tokenizer code (in case some of the expressions might have caused the problem):
```
# modify prefix pattern
split_chars = lambda char: list(char.strip().split(" "))
prefixes = (
                ["§", "%", "=", "—", "–", r"\+(?![0-9])"]
                + split_chars( r"… …… , : ; \! \? ¿ ؟ ¡ \( \) \[ \] \{ \}  _ # \ & 。 ？ ！ ， 、 ； ： ～ · । ، ۔ ؛ ٪") #removed [*<>]
                + LIST_ELLIPSES
                + LIST_QUOTES
                + LIST_CURRENCY
                + LIST_ICONS
            )

prefix_re = compile_prefix_regex(prefixes)

# Modify tokenizer infix patterns
infixes = (
            LIST_ELLIPSES
            + LIST_ICONS
            + [
                r"(?<=[0-9])[+\-\*^](?=[0-9-])",
                r"(?<=[{al}{q}])\.(?=[{au}{q}])".format(
                    al=ALPHA_LOWER, au=ALPHA_UPPER, q=CONCAT_QUOTES
                ),
                r"(?<=[{a}]),(?=[{a}])".format(a=ALPHA),
                r"(?<=[{a}])(?:{h})(?=[{a}])".format(a=ALPHA, h=HYPHENS),
                r"(?<=[{a}0-9])[:<>=/](?=[{a}])".format(a=ALPHA),
                r"(?<=[{a}0-9])[:;()/=](?=[{a}0-9])".format(a=ALPHA),  # added this to separate 'obs:30'
                r"(?<=[{a}0-9])[()][:;](?=[{a}0-9])".format(a=ALPHA),  # added this to separate '(obs):30'
                r"(?<=[{a}])(?=[0-9])".format(a=ALPHA),             # to separate something123
                r"(?<=[0-9])(?=[{a}])".format(a=ALPHA),             # to separate 3mmHg
            ]
        )
infix_re = compile_infix_regex(infixes)

# Modify suffix pattern
suffixes = (
            LIST_PUNCT
            + LIST_ELLIPSES
            + LIST_QUOTES
            + LIST_ICONS
            + ["'s", "'S", "’s", "’S", "—", "–"]
            + [
                r"(?<=[0-9])\+",
                r"(?<=°[FfCcKk])\.",
                r"(?<=[0-9])(?:{c})".format(c=CURRENCY),
                r"(?<=[0-9])(?:{u})".format(u=UNITS),
                r"(?<=[0-9{al}{e}{p}(?:{q})])\.".format(
                    al=ALPHA_LOWER, e=r"%²\-\+", q=CONCAT_QUOTES, p=PUNCT
                ),
                r"(?<=[{au}][{au}])\.".format(au=ALPHA_UPPER),
            ]
        )
suffix_re = compile_suffix_regex(suffixes)

nlp.tokenizer.prefix_search = prefix_re.search
nlp.tokenizer.suffix_search = suffix_re.search
nlp.tokenizer.infix_finditer = infix_re.finditer
```

**Comment by adrianeboyd** [comment, marked as answer]
To debug tokenizer settings (it can definitely be tricky!), use `nlp.tokenizer.explain(text)` to show which regex pattern is matching which part of the token.

(Edited to add: there is a very minor known bug related to prefixes in suffix lookbehinds, but it only matters if you notice that you get different tokens from `nlp.tokenizer(text)` and `nlp.tokenizer.explain(text)`.)

**Reply by farrandi** [reply]
Thanks for the reply.

I tried using the function:
```
text = "abc:1"
doc = nlp(text)
print(nlp.tokenizer.explain(text))
```

But I only got:
`[('TOKEN', 'abc'), ('SPECIAL-1', ':1')]`

I am not sure what 'SPECIAL-1' is supposed to be.

**Reply by adrianeboyd** [reply]
The names correspond to the tokenizer settings described in the following links: https://spacy.io/api/tokenizer#init, https://spacy.io/usage/linguistic-features#tokenization

`SPECIAL-1` is the first token in a special case, which are also sometimes called "exceptions" or "rules".

This is probably from a kind of out-dated list of emoticons that have been in spacy's defaults for a long time. You can remove individual exceptions by modifying `nlp.tokenizer.rules` as shown here: https://github.com/explosion/spaCy/discussions/8218#discussioncomment-838238

(If you're editing `nlp.tokenizer.rules`, assign a modified dict rather than editing `nlp.tokenizer.rules` directly because otherwise you may run into confusion with the tokenizer cache.)
